### PR TITLE
KAFKA-7223: Suppression Buffer Metrics

### PR DESCRIPTION
**Pull request by vvcephei**
Add the final batch of metrics from KIP-328

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+import static org.apache.kafka.common.serialization.Serdes.String;
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.apache.kafka.streams.kstream.Suppressed.BufferConfig.maxRecords;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.core.Is.is;
+
+@SuppressWarnings("PointlessArithmeticExpression")
+public class KTableSuppressProcessorMetricsTest {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java:44]: I did the metrics in a separate file to keep related tests together for readability.

I'm aware there are a lot of repeated strings in the metric names, but it is on purpose: it's the only place in the code where you can visually confirm that the metrics we're measuring conform to the ones we document.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java [hunk 2]
```diff
@@ -20,12 +20,19 @@
...
+import org.apache.kafka.common.metrics.stats.Total;
 import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
 import org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl;
 
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
 
 public class Sensors {
+
+    private static final String PROCESSOR_NODE_METRICS_GROUP = "stream-processor-node-metrics";
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: might as well consolidate this repeated string.

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @mjsax @bbejeck ,

Do you mind taking a look at this last PR for the KIP?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java [hunk 3]
```diff
@@ -38,13 +45,89 @@ public static Sensor lateRecordDropSensor(final InternalProcessorContext context
...
+                "The total number of occurrence of " + operation + " operations.",
+                tags
+            ),
+            new Total()
+        );
+
+        return sensor;
+    }
+
+    public static Sensor suppressionEmitSensor(final InternalProcessorContext context) {
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java:86]: This sensor is different than that proposed in the kip. During implementation, I noticed a weird asymmetry in which the processor node would measure the suppression, but the buffer would measure the "eviction" aka "emission" aka "forwarding".

I'm proposing to update the KIP to make these two measurements symmetrical.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java [hunk 4]
```diff
@@ -287,8 +295,13 @@ public void evictWhile(final Supplier<Boolean> predicate,
                     next = null;
                     minTimestamp = Long.MAX_VALUE;
                 }
+
+                evictions++;
             }
         }
+        if (evictions > 0) {
+            updateBufferMetrics();
+        }
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:320]: Avoid recording metrics if they haven't changed.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java [hunk 5]
```diff
@@ -44,5 +51,67 @@ public static Sensor createTaskAndStoreLatencyAndThroughputSensors(final Sensor.
...
+        final String sensorName = "suppression-buffer-" + property;
+
+        final Sensor sensor = metrics.storeLevelSensor(
+            context.taskId().toString(),
+            store.name(),
+            sensorName,
+            Sensor.RecordingLevel.DEBUG
+        );
+
+        final String metricsGroup = "stream-buffer-metrics";
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:77]: In the KIP, I erroneously proposed to make this a node-level sensor. It should be a store-level sensor instead. I'm proposing to update the KIP if you all agree.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressScenarioTest.java [hunk 6]
```diff
@@ -368,7 +368,7 @@ public void shouldSupportFinalResultsForTimeWindows() {
         final KTable<Windowed<String>, Long> valueCounts = builder
             .stream("input", Consumed.with(STRING_SERDE, STRING_SERDE))
             .groupBy((String k, String v) -> k, Grouped.with(STRING_SERDE, STRING_SERDE))
-            .windowedBy(TimeWindows.of(2L).grace(ofMillis(1L)))
+            .windowedBy(TimeWindows.of(ofMillis(2L)).grace(ofMillis(1L)))
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressScenarioTest.java]: just cleaning up deprecated usages.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java [hunk 7]
```diff
@@ -77,9 +77,16 @@
...
-                new KTableSuppressProcessor<>(getImpl(suppressed), storeName, keySerde, new FullChangeSerde<>(valueSerde));
+                new KTableSuppressProcessor<>(
+                    (SuppressedInternal<K>) suppressed,
+                    storeName,
+                    keySerde,
+                    new FullChangeSerde<>(valueSerde)
+                );
 
             final MockInternalProcessorContext context = new MockInternalProcessorContext();
+            context.setCurrentNode(new ProcessorNode("testNode"));
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java:88]: need this for the node-level metrics

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java [hunk 8]
```diff
@@ -461,10 +468,6 @@ public boolean matches(final Object item) {
         };
     }
 
-    private static <K> SuppressedInternal<K> getImpl(final Suppressed<K> suppressed) {
-        return (SuppressedInternal<K>) suppressed;
-    }
-
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java:467]: inline unnecessary method. I must have made this when I thought that the cast would cause a compiler warning, but it does not.

File: streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java [hunk 9]
```diff
@@ -208,7 +210,12 @@ public MockProcessorContext(final Properties config, final TaskId taskId, final
         this.taskId = taskId;
         this.config = streamsConfig;
         this.stateDir = stateDir;
-        this.metrics = new StreamsMetricsImpl(new Metrics(), "mock-processor-context-virtual-thread");
+        final MetricConfig metricConfig = new MetricConfig();
+        metricConfig.recordLevel(Sensor.RecordingLevel.DEBUG);
```
Comment (vvcephei) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java:214]: I think it should be fine to set the recording level to debug for all unit tests, rather than create a way to configure it.

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @mjsax @bbejeck 

I've updated this PR, and it's ready for review, now that 2.1 is complete.

These metrics would be included in the 2.2 release.

**Comment by vvcephei** [comment]
java 11 had six unrelated failures:
* kafka.api.ClientIdQuotaTest.testThrottledProducerConsumer
* kafka.api.UserClientIdQuotaTest.testThrottledProducerConsumer
* kafka.api.UserQuotaTest.testThrottledProducerConsumer
* kafka.api.UserQuotaTest.testQuotaOverrideDelete
* org.apache.kafka.common.network.SslTransportLayerTest.testListenerConfigOverride
* kafka.tools.MirrorMakerIntegrationTest.testCommaSeparatedRegex

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java [hunk 10]
```diff
@@ -38,13 +45,89 @@ public static Sensor lateRecordDropSensor(final InternalProcessorContext context
...
+        final Map<String, String> tags = metrics.tagMap(
+            "task-id", context.taskId().toString(),
+            "processor-node-id", context.currentNode().name()
+        );
+
+        sensor.add(
+            new MetricName(
+                operation + "-rate",
+                PROCESSOR_NODE_METRICS_GROUP,
+                "The average number of occurrence of " + operation + " operation per second.",
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: nit: Should the string say `The rate of...`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java [hunk 11]
```diff
@@ -38,13 +45,89 @@ public static Sensor lateRecordDropSensor(final InternalProcessorContext context
...
+            Sensor.RecordingLevel.DEBUG
+        );
+
+        final Map<String, String> tags = metrics.tagMap(
+            "task-id", context.taskId().toString(),
+            "processor-node-id", context.currentNode().name()
+        );
+
+        sensor.add(
+            new MetricName(
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java:102]: seems like the contents of the two `sensor#add` methods are the same as above maybe refactor into two methods?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java (hunk 5 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:77]: Yeah, that makes sense to me. +1

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        processor.process(key, value);
+
+        {
+            final Map<MetricName, ? extends Metric> metrics = context.metrics().metrics();
+
+            assertThat(metrics.get(EVICTION_RATE_METRIC).metricValue(), is(0.0));
+            assertThat(metrics.get(EVICTION_TOTAL_METRIC).metricValue(), is(0.0));
+            assertThat((Double) metrics.get(SUPPRESSION_RATE_METRIC).metricValue(), greaterThan(0.0));
+            assertThat(metrics.get(SUPPRESSION_TOTAL_METRIC).metricValue(), is(1.0));
+            assertThat(metrics.get(BUFFER_SIZE_AVG_METRIC).metricValue(), is(51.0));
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java]: I feel like I should know the answer to this, but how do we `51` for the answer here and below?

File: streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java (hunk 9 above)
Comment (bbejeck) [reply streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java:214]: I agree

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 3 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java:86]: I didn't think of that before, but now that you mention it, the change makes sense to me.

**Review by bbejeck — COMMENTED** [review]
I made a pass over the PR and left some comments, overall looks good.

**Comment by bbejeck** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: sure!

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java (hunk 12 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java]: Empirically :/

It's just the result of summing all the pieces of the test data.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: Actually, strange as the sentence is, this is what the other rate metrics say, so I think we should just keep it for consistency.

**Comment by vvcephei** [comment]
Hi @bbejeck , Thanks for the review!

As explained above, that description string for the rate metrics is common across the other rate metrics, so I've left it as-is. Is that ok?

Incidentally, I was surprised that the tests didn't fail when I changed the description, since I specifically tested for them. Turns out, MetricName equality is defined to ignore the description.

I've refactored the test to verify the description as well, and also fixed a problem with one of the description strings I found.

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by bbejeck** [comment]
retest this please

**Comment by vvcephei** [comment]
rebased

**Comment by vvcephei** [comment]
Java 8 failure unrelated.
* ConsumerBounceTest.testClose: created https://issues.apache.org/jira/browse/KAFKA-7540

Java 11 failures unrelated:
* DynamicBrokerReconfigurationTest.testUncleanLeaderElectionEnable: created https://issues.apache.org/jira/browse/KAFKA-7541
* GssapiAuthenticationTest.testServerAuthenticationFailure: created https://issues.apache.org/jira/browse/KAFKA-7542
* EosIntegrationTest.shouldNotViolateEosIfOneTaskFails: created https://issues.apache.org/jira/browse/KAFKA-7544

Retest this, please.

**Comment by vvcephei** [comment]
hmm. It looks like the java 11 build hung and then timed out. Here's the end of the output:
```
...
00:46:27.275 kafka.server.ServerGenerateBrokerIdTest > testConsistentBrokerIdFromUserConfigAndMetaProps STARTED
00:46:29.775 
00:46:29.775 kafka.server.ServerGenerateBrokerIdTest > testConsistentBrokerIdFromUserConfigAndMetaProps PASSED
03:00:51.124 Build timed out (after 180 minutes). Marking the build as aborted.
03:00:51.440 Build was aborted
03:00:51.492 [FINDBUGS] Skipping publisher since build result is ABORTED
03:00:51.492 Recording test results
03:00:51.495 Setting GRADLE_4_8_1_HOME=/home/jenkins/tools/gradle/4.8.1
03:00:58.017 Setting GRADLE_4_8_1_HOME=/home/jenkins/tools/gradle/4.8.1
03:00:59.330 Setting GRADLE_4_8_1_HOME=/home/jenkins/tools/gradle/4.8.1
03:00:59.331 Adding one-line test results to commit status...
03:00:59.332 Setting GRADLE_4_8_1_HOME=/home/jenkins/tools/gradle/4.8.1
03:00:59.334 Setting GRADLE_4_8_1_HOME=/home/jenkins/tools/gradle/4.8.1
03:00:59.335 Setting status of 5bdcd0e023c6f406d585155399f6541bb6a9f9c2 to FAILURE with url https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/266/ and message: 'FAILURE
03:00:59.335  9053 tests run, 1 skipped, 0 failed.'
03:00:59.335 Using context: JDK 11 and Scala 2.12
03:00:59.541 Setting GRADLE_4_8_1_HOME=/home/jenkins/tools/gradle/4.8.1
03:00:59.542 Finished: ABORTED
```

I reported this as https://issues.apache.org/jira/browse/KAFKA-7553

**Comment by vvcephei** [comment]
Retest this, please.

**Comment by vvcephei** [comment]
The tests failed again, but this time, I didn't catch what failed. I'll take the opportunity to rebase, which will run the tests again.

**Comment by vvcephei** [comment]
Java 8 passed, and Java 11 hung again.
I added a comment to https://issues.apache.org/jira/browse/KAFKA-7553

Retest this, please.

**Comment by vvcephei** [comment]
Aha! The tests finally passed.

@mjsax or @guozhangwang , do you mind taking a look?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java [hunk 13]
```diff
@@ -77,6 +84,7 @@ public void init(final ProcessorContext context) {
     public void process(final K key, final Change<V> value) {
         buffer(key, value);
         enforceConstraints();
+        suppressionSensor.record();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java]: This seems to effectively count the number of input records (total and average) -- so it's basically input data rate? Is it intended like this. `intermediate-result-suppression` sounds more like how often something is suppressed, but I don't see how this would be reflected here.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java [hunk 14]
```diff
@@ -355,4 +372,9 @@ private long computeRecordSize(final Bytes key, final ContextualRecord value) {
         }
         return size;
     }
+
+    private void updateBufferMetrics() {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:378]: Should we call this in `init()` and/or `close()`, too?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: I am always confused by metric. What it the group name used for (ie, semantic meaning)?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java [hunk 15]
```diff
@@ -25,7 +31,8 @@
 import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.addInvocationRateAndCount;
 
 public final class Sensors {
-    
+    private static final String PROCESSOR_NODE_METRICS_GROUP = "stream-processor-node-metrics";
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java]: Does it make sense to use the same group name as above?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java (hunk 13 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java]: I struggled a bit with this question as well. Technically, everything that goes into the buffer is suppressed, and then some stuff is emitted later. We have one metric for each of these things.

I agree it seems more like we should be measuring the number of records that are discarded. I guess we could do this by recording the metric when we buffer a record for which there was a pre-existing record in the buffer. This would be straightforward for the in-memory buffer, but I'm concerned about doing an unnecessary `get` with an on-disk store to maintain the metric.

You should be able to arrive at the same (or similar) number by subtracting the emit metric from the suppress metric. I was thinking this is what people would probably do, but on second though, I could put together a composite metrics that does it internally.

WDYT?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java (hunk 15 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java]: This string can effectively never change (it would mean we've renamed all our processor metrics), so I don't think there's any risk from the duplication.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 2 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: AFAICT, I don't think it's "used for" anything in particular. It's part of the metric name, but so is the description.

I assume it's intended to be like a namespace.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: group name is used as a first "tag" of the metric name in JMX reporter: `xxx-metrics:type=[group-name],[tag1]=[value1],...`; for other reporters they can use the group name however they like.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java]: I'd prefer only providing the non-composite metrics out-of-box while letting users to compose whenever they want, this is along with the same argument @vvcephei had about "always recording at the lowest level granularity only, and let users to the roll-up themselves if they like".

Following this line, I'd suggest we only keep one metrics aka `suppressionEmitSensor` and users can then calculate the suppression rate by "processing-rate" minus "suppression-emit-rate".

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:77]: +1

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:77]: This is not newly introduced in this PR but: although we call it `stream-buffer-metrics` where the `buffer-id` is the store-name, BUT the store name could be an auto-generated name. Note for the named cache we call it `stream-record-cache-metrics` but the cache name is the store name, and hence we have the same issue.

But here we use the store.name which would be `KTABLE-SUPPRESS-XXXX-store` if it is auto generated, maybe it's less confusing to use the processor-name instead which would be `KTABLE-SUPPRESS-XXX` if it is auto generated. Note even if `Suppressed.name()` is specified as `XYZ` with store.name() we still set the buffer-id as `XYZ-store` instead.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java [hunk 16]
```diff
@@ -57,6 +60,8 @@
     private long minTimestamp = Long.MAX_VALUE;
     private RecordCollector collector;
     private String changelogTopic;
+    private Sensor bufferSizeSensor;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:63]: In the wiki there is another metric for `suppression-mem-buffer-evict`, is that not in the scope of this PR?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        context.setStreamTime(timestamp + 1);
+        context.setRecordMetadata("", 0, 1L, null, timestamp + 1);
+        processor.process("key", value);
+
+        {
+            final Map<MetricName, ? extends Metric> metrics = context.metrics().metrics();
+
+            verifyMetric(metrics, EVICTION_RATE_METRIC, greaterThan(0.0));
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java:184]: Is this rate value be time-dependent (I remember we saw some issues with it if the test runs too slowly), to be non-zero in some edge cases?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java (hunk 13 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java]: Ok, I can get behind this. It should be measuring the same thing as the `process-rate` metric for the suppression ProcessorNode.

I'll remove this metric.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java (hunk 14 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:378]: sure; I don't think they're strictly necessary, but it doesn't hurt, and it'll make the system more resilient to future changes.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java (hunk 16 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:63]: Ah, now that you mention it, when I originally created this PR, I noticed there were some shortcomings with the metrics I proposed in the KIP, so I wanted to get the reviewers' feedback on the form it has taken in this PR before proposing an update to the KIP.

In particular, I proposed an asymmetric pair of metrics where the processor node would measure the number of incoming events (`intermediate-result-suppression`), but the buffer would measure the outgoing (emitted==evicted) events. It seemed better to offer a symmetric pair of in/out metrics on either the processor or the buffer. In this PR, I chose to put them both on the node.

In other words, I'm proposing to replace `suppression-mem-buffer-evict` with `suppression-emit`.

Of course, in this same review, you have proposed to drop `intermediate-result-suppression` in favor of the existing `process-rate` and `process-total` node-level metrics. This reduces the argument, but I *think* it still makes sense to make this a node metric instead of a buffer metric, since (IMHO) it more closely matches the intent of using a suppression *node* to control the emission rate.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java (hunk 17 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorMetricsTest.java:184]: IIRC, rate metrics are time-dependent, but I think that this particular comparison is pretty safe for a mock processor unit test.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java (hunk 5 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java:77]: I think it's good to be aware of this, but it seems like reporting the store name is the right thing to do here.

Note that the buffer in this case *is* the store for the suppression processor. So, using the processor name instead of the full store name would be more analogous to doing the same thing for a K/V Store.

As written, the metrics that are specific to the buffer (aka store) will be tagged with a name that matches the name of that component when you describe the topology.

Does that seem right, or have I missed your point?

**Comment by vvcephei** [comment]
Hi @mjsax @guozhangwang @bbejeck ,

Thank you all for your reviews!

I have updated this PR to address the outstanding comments. If it looks good to you now, I will go ahead and update the KIP (and send out a notice to the vote thread), and we can merge the PR.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java (hunk 16 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:63]: Thanks for the clarification, that sounds good to me.

**Comment by guozhangwang** [comment]
LGTM! @mjsax please feel free to merge if you think all your comments are addressed.

**Comment by vvcephei** [comment]
KIP is updated and message sent to the vote thread.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java (hunk 15 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/Sensors.java]: It's not about any risk, but a semantic question. Is it correct to use the same name (ie, right hand side) for both? And if yes, why so we want this?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: Sorry for my confusion. Let me rephrase my current understanding:
A "metric" is a single value that we track and report as key-value pair. A Sensor groups multiple metric together for ease of use -- each metric gets globally unique name that is put together with many parts.

1.  We prefix metric names (the `xxx-metrics:` part) with the `clientId` that is different for each `StreamsThread`.
2. What is the purpose of `group-name` in the metric-name, ie, what metrics should use the same or different group name 
3. Each sensor has a name that is added to all metric names within the sensor (ie, the sensor name groups all it's contains metrics)
4. we also put additional tags to add more meta information (task-id, processor-id) if appropriate to name each metric uniquely within a sensor.

To clarify: by "grouping" I meant use the same string for a specific part in the metric name. Ie, the prefix groups all metric based on the stream-thread, and the sensor name groups all it's contained metrics (as sub-group within the stream-thread group of metrics). Does this make sense?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: To clarify: there are two entities: the metrics registry which organizes the metrics, and there is metrics reporter which regularly pulls from the registry to report the metric values. 

Inside metrics registry there is `sensor` which as you understand is just a way of grouping metrics into meaningful clusters. The `SensorName` is just an id for distinguishing sensors in the metrics registry (i.e. you will see logic like `if this sensor as already been created in the registry skip this step)`. A metric name presented in `MetricName` which contains groupName, tags, etc is just a logical entity in the registry. How to represent the metric names is up to the metrics reporter (different reporters can definitely represent it differently). As for the sensor names, they should never be seen outside the registry as the metrics reporter never exposed them.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: Thanks @guozhangwang 

I was just checking, and we also define `"stream-processor-node-metrics"` in `ProcessorNode` -- should we unify both, to have one constant only?

Think, we can also simplify `ProcessorNode#createTaskAndNodeLatencyAndThroughputSensors` and remove the `group` parameter.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java [hunk 18]
```diff
@@ -189,12 +199,13 @@ public boolean isOpen() {
 
     @Override
     public void close() {
+        open = false;
         index.clear();
         sortedMap.clear();
         dirtyKeys.clear();
         memBufferSize = 0;
         minTimestamp = Long.MAX_VALUE;
-        open = false;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:197]: Why moving this?

**Review by mjsax — COMMENTED** [review]
Two more follow up question. Overall LGTM.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java (hunk 18 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:197]: so that the buffer will report as "closed" immediately after this method is called, rather than only after the close operation has completed.

In retrospect, it seemed safer for :
* `open()` to set `open=true` at when opening has *completed*
* `close()` to set `open=false` when closing has *started*

This way, observers who first check that a store is open before querying would never be fooled into querying a partially closed store.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java (hunk 2 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: Actually, someone filed a Jira last week reporting that a heap analysis identified this exact group name as responsible for Megabytes of heap space, so perhaps we should consolidate it into one constant!

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: If this is an issue, maybe we should do a fix first (in it's own PR, so we can cherry-pick), and rebase/merge this PR later? Thought? Are you planning to do a PR for the reported issue?

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: oh. I've already added a commit to this PR.

I am planning a separate PR for a couple of other issues he identified. It's https://issues.apache.org/jira/browse/KAFKA-7660, by the way.

**Comment by vvcephei** [comment]
Unrelated test failure:
```
kafka.api.CustomQuotaCallbackTest.testCustomQuotaCallback

Error Message
java.lang.AssertionError: Partition [group1_largeTopic,69] metadata not propagated after 15000 ms
Stacktrace
java.lang.AssertionError: Partition [group1_largeTopic,69] metadata not propagated after 15000 ms
	at kafka.utils.TestUtils$.fail(TestUtils.scala:351)
	at kafka.utils.TestUtils$.waitUntilTrue(TestUtils.scala:761)
	at kafka.utils.TestUtils$.waitUntilMetadataIsPropagated(TestUtils.scala:850)
	at kafka.utils.TestUtils$$anonfun$createTopic$2.apply(TestUtils.scala:330)
	at kafka.utils.TestUtils$$anonfun$createTopic$2.apply(TestUtils.scala:329)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at scala.collection.MapLike$DefaultKeySet.foreach(MapLike.scala:174)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:47)
	at scala.collection.SetLike$class.map(SetLike.scala:92)
	at scala.collection.AbstractSet.map(Set.scala:47)
	at kafka.utils.TestUtils$.createTopic(TestUtils.scala:329)
	at kafka.utils.TestUtils$.createTopic(TestUtils.scala:312)
	at kafka.api.CustomQuotaCallbackTest.createTopic(CustomQuotaCallbackTest.scala:180)
	at kafka.api.CustomQuotaCallbackTest.testCustomQuotaCallback(CustomQuotaCallbackTest.scala:135)
```

Retest this, please

**Comment by vvcephei** [comment]
@mjsax , do you have any more feedback on this?

Thanks,
-John

**Comment by guozhangwang** [comment]
> @mjsax , do you have any more feedback on this?

I think there is none as he said "otherwise LGTM" :) Will merge to trunk now.

**Comment by guozhangwang** [comment]
> I was just checking, and we also define "stream-processor-node-metrics" in ProcessorNode -- should we unify both, to have one constant only?

> Think, we can also simplify ProcessorNode#createTaskAndNodeLatencyAndThroughputSensors and remove the group parameter.

There are a bunch of those constants that appears on different classes and I remember @vvcephei mentioned the reason not doing it in this PR is to not introduce mutual dependency on those classes. If we want to do so I'd suggest we just exact all these constants into a separate class and let all other sensor classes refer to this single one, which we can do in another minor PR.

**Comment by guozhangwang** [comment]
> There are a bunch of those...

My bad, I did not refresh on the file changes and hence was looking at an older version. The above comment is actually already addressed.
